services/web/server/src/simcore_service_webserver/meta_iterations.py

```diff
@@ -0,0 +1,332 @@
+    # NOTE: prefers to a json dump because it allows regex search?
+    # should be searchable e.g. get repo/*/*/*
+    # can do the same with json since it is a str!?
+    return (
+        f"iteration:{repo_commit_id}/{iter_index}/{total_count}/{parameters_checksum}"
+    )
+
+
+def parse_iteration_tag_name(name: str) -> Dict[str, Any]:
+    if m := re.match(
+        r"^iteration:(?P<repo_commit_id>\d+)/(?P<iter_index>\d+)/(?P<total_count>-*\d+)/(?P<parameters_checksum>.*)$",
+        name,
+    ):
+        return m.groupdict()
+    return {}
+
+
+# GET/CREATE iterations ------------------------------------------------------------
+
+
+async def get_or_create_runnable_projects(
+    request: web.Request,
+    project_uuid: ProjectID,
+) -> Tuple[List[ProjectID], List[CommitID]]:
+    """
+    Returns ids and refid of projects that can run
+    If project_uuid is a std-project, then it returns itself
+    If project_uuid is a meta-project, then it returns iterations
+    """
+
+    # FIXME: what happens if not
+
+    vc_repo = VersionControlForMetaModeling(request)
+    assert vc_repo.user_id  # nosec
+
+    # TODO:  handle UserUndefined and translate into web.HTTPForbidden
+    project: ProjectDict = await vc_repo.get_project(str(project_uuid))
+
+    project_nodes: Dict[NodeID, Node] = {
+        nid: Node.parse_obj(n) for nid, n in project["workbench"].items()
+    }
+
+    # init returns
+    runnable_project_vc_commits: List[CommitID] = []
+    runnable_project_ids: List[ProjectID] = [
+        project_uuid,
+    ]
+
+    # auto-commit
+    #   because it will run in parallel -> needs an independent working copy
+    repo_id = await vc_repo.get_repo_id(project_uuid)
+    if repo_id is None:
+        repo_id = await vc_repo.init_repo(project_uuid)
+
+    main_commit_id = await vc_repo.commit(
+        repo_id,
+        tag=f"auto:main/{project_uuid}",
+        message="auto-commit at get_or_create_runnable_projects",
+    )
+    runnable_project_vc_commits.append(main_commit_id)
+
+    # std-project
+    is_meta_project = any(
+        is_iterator_service(node.key) for node in project_nodes.values()
+    )
+    if not is_meta_project:
+        return runnable_project_ids, runnable_project_vc_commits
+
+    # meta-project: resolve project iterations
+    runnable_project_ids = []
+    runnable_project_vc_commits = []
+
+    iterations = _build_project_iterations(project_nodes)
+    log.debug(
+        "Project %s with %s parameters, produced %s variants",
+        project_uuid,
+        len(iterations[0]) if iterations else 0,
+        len(iterations),
+    )
+
+    # Each iteration generates a set of 'parameters'
+    #  - parameters are set in the corresponding outputs of the meta-nodes
+    #
+    parameters: Parameters
+    updated_nodes: NodesDict
+    total_count = len(iterations)
+    original_name = project["name"]
+
+    for iter_index, (parameters, updated_nodes) in enumerate(iterations):
```

Comment:
what is the use of the parenthesis?

Reply:
that is standard python :-)
this is how you can unzip the iterator while looping. 

It is a compact way to write this:
```python
for iter_index, value in enumerate(iterations):
      parameters, updated_nodes = value
       ...
```